lib/srv/app/gcp/handler.go

```diff
@@ -0,0 +1,276 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	recorder := httplib.NewResponseStatusRecorder(w)
+	s.fwd.ServeHTTP(recorder, fwdRequest)
+	status := uint32(recorder.Status())
+
+	if err := sessionCtx.Audit.OnRequest(req.Context(), sessionCtx, fwdRequest, status, nil); err != nil {
+		// log but don't return the error, because we already handed off request/response handling to the oxy forwarder.
+		s.Log.WithError(err).Warn("Failed to emit audit event.")
+	}
+	return nil
+}
+
+func (s *handler) formatForwardResponseError(rw http.ResponseWriter, r *http.Request, err error) {
+	s.Log.WithError(err).Debugf("Failed to process request.")
+	common.SetTeleportAPIErrorHeader(rw, err)
+
+	// Convert trace error type to HTTP and write response.
+	code := trace.ErrorToCode(err)
+	http.Error(rw, http.StatusText(code), code)
+}
+
+// prepareForwardRequest prepares a request for forwarding, updating headers and target host. Several checks are made along the way.
+func (s *handler) prepareForwardRequest(r *http.Request, sessionCtx *common.SessionContext) (*http.Request, error) {
+	forwardedHost := r.Header.Get("X-Forwarded-Host")
+	if !gcp.IsGCPEndpoint(forwardedHost) {
+		return nil, trace.AccessDenied("%q is not a GCP endpoint", forwardedHost)
+	}
+
+	payload, err := awsutils.GetAndReplaceReqBody(r)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	reqCopy, err := http.NewRequest(r.Method, r.URL.String(), bytes.NewReader(payload))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	reqCopy.URL.Scheme = "https"
+	reqCopy.URL.Host = forwardedHost
+	reqCopy.Header = r.Header.Clone()
+
+	err = s.replaceAuthHeaders(r, sessionCtx, reqCopy)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return reqCopy, trace.Wrap(err)
+}
+
+func (s *handler) replaceAuthHeaders(r *http.Request, sessionCtx *common.SessionContext, reqCopy *http.Request) error {
+	auth := reqCopy.Header.Get("Authorization")
+	if auth == "" {
+		s.Log.Debugf("No Authorization header present, skipping replacement.")
+		return nil
+	}
+
+	s.Log.Debugf("Processing request, sessionId = %q, gcpServiceAccount = %q", sessionCtx.Identity.RouteToApp.SessionID, sessionCtx.Identity.RouteToApp.GCPServiceAccount)
+	token, err := s.getToken(r.Context(), sessionCtx.Identity.RouteToApp.GCPServiceAccount)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Set new authorization
+	reqCopy.Header.Set("Authorization", "Bearer "+token.AccessToken)
+	return nil
+}
+
+type cacheKey struct {
+	serviceAccount string
+}
+
+const getTokenTimeout = time.Second * 5
+
+// the scope list if fixed for now, but could be either be extended or made customizable
+// if an important use case were to be presented.
+var defaultScopeList = []string{
+	"https://www.googleapis.com/auth/cloud-platform",
+
+	"openid",
+	"https://www.googleapis.com/auth/userinfo.email",
+	"https://www.googleapis.com/auth/appengine.admin",
+	"https://www.googleapis.com/auth/sqlservice.login",
+	"https://www.googleapis.com/auth/compute",
```

Comment:
Could you add the docs for the GCP scopes that we are using ?

Reply:
Done.